share/sysbuild/cmake/modules/sysbuild_extensions.cmake

```diff
@@ -107,6 +107,87 @@ function(sysbuild_get variable)
+#                     The sysbuild cahe is only updated if it contain changes.
+# APPLICATION <name>: Name of the application.
+# CMAKE_RERUN       : Force a CMake rerun for the application during next build
+#                     invocation if the sysbuild cache has changed. It is
+#                     advised to always use this flag. Not using this flag can
+#                     reduce build time, but only do so if application is
+#                     guranteed to be up-to-date.
+#
+function(sysbuild_cache)
+  cmake_parse_arguments(SB_CACHE "CREATE;CMAKE_RERUN" "APPLICATION" "" ${ARGN})
+  zephyr_check_arguments_required(sysbuild_cache SB_CACHE APPLICATION)
+  zephyr_check_flags_required(sysbuild_cache SB_CACHE CREATE )
+
+  get_cmake_property(sysbuild_cache CACHE_VARIABLES)
+
+  foreach(var_name ${sysbuild_cache})
+    if(NOT "${var_name}" MATCHES "^(CMAKE_.*|BOARD)$")
+      # Perform a dummy read to prevent a false warning about unused variables
+      # being emitted due to a cmake bug: https://gitlab.kitware.com/cmake/cmake/-/issues/24555
+      set(unused_tmp_var ${${var_name}})
+
+      # We don't want to pass internal CMake variables.
+      # Required CMake variable to be passed, like CMAKE_BUILD_TYPE must be
+      # passed using `-D` on command invocation.
+      get_property(var_type CACHE ${var_name} PROPERTY TYPE)
+      set(cache_entry "${var_name}:${var_type}=$CACHE{${var_name}}")
+      string(REPLACE ";" "\;" cache_entry "${cache_entry}")
+      list(APPEND sysbuild_cache_strings "${cache_entry}\n")
+    endif()
+  endforeach()
+  if(DEFINED BOARD_REVISION)
+    list(APPEND sysbuild_cache_strings "BOARD:STRING=${BOARD}@${BOARD_REVISION}\n")
+  else()
+    list(APPEND sysbuild_cache_strings "BOARD:STRING=${BOARD}\n")
+  endif()
+  list(APPEND sysbuild_cache_strings "SYSBUILD_NAME:STRING=${SB_CACHE_APPLICATION}\n")
+
+  if(${SB_CACHE_APPLICATION}_MAIN_APP)
+    list(APPEND sysbuild_cache_strings "SYSBUILD_MAIN_APP:BOOL=True\n")
+  endif()
+
+  if(${SB_CACHE_APPLICATION}_BOARD AND NOT DEFINED CACHE{${SB_CACHE_APPLICATION}_BOARD})
+    # Only set image specific board if provided.
+    # The sysbuild BOARD is exported through sysbuild cache, and will be used
+    # unless <image>_BOARD is defined.
+    list(APPEND sysbuild_cache_strings
+         "${SB_CACHE_APPLICATION}_BOARD:STRING=${${SB_CACHE_APPLICATION}_BOARD}\n"
+    )
+  endif()
+
+  get_target_property(${SB_CACHE_APPLICATION}_CACHE_FILE ${SB_CACHE_APPLICATION} CACHE_FILE)
+  print(${SB_CACHE_APPLICATION}_CACHE_FILE)
+  file(WRITE ${${SB_CACHE_APPLICATION}_CACHE_FILE}.tmp ${sysbuild_cache_strings})
+  if(SB_CACHE_CMAKE_RERUN)
+    execute_process(COMMAND ${CMAKE_COMMAND} -E compare_files
+                    ${${SB_CACHE_APPLICATION}_CACHE_FILE}.tmp
+                    ${${SB_CACHE_APPLICATION}_CACHE_FILE}
+                    RESULT_VARIABLE compare_res
+    )
+    if(NOT compare_res EQUAL 0)
+      file(COPY_FILE ${${SB_CACHE_APPLICATION}_CACHE_FILE}.tmp
```

Comment:
Shouldn't this use `zephyr_file_copy()` still?

Reply:
no, only reason for having `zephyr_file_copy()` is for the `ONLY_IF_DIFFERENT` argument, which we aren't using here.

We just checked a couple of lines above if the files were different, and the are, so then we should just copy.

Reply:
But `zephyr_file_copy()` is described as a stand-in for `file(COPY_FILE)`, which wasn't introduced until CMake 3.21. It's not just the optional `ONLY_IF_DIFFERENT` argument that's new in that version. See https://cmake.org/cmake/help/latest/command/file.html#copy-file

If we can use `file(COPY_FILE)` here, then we can also use `file(COPY_FILE ... ONLY_IF_DIFFERENT)` below.

Reply:
> which wasn't introduced until CMake 3.21

My bad, should have used `COPY`, not `FILE_COPY` (which has the `ONLY_IF_DIFFERENT` signature).

Thanks for spotting, will change.

Reply:
to `zephyr_file_copy()`, as I need to change file name in the process.

Reply:
fixed, thanks for spotting